### PR TITLE
ldelf: scripts/gen_ldelf_hex.py: allow several non-writable LOAD segments

### PR DESCRIPTION
**Pull request by jforissier**
```
Latest Clang [1] generates the following ldelf.elf (QEMU):

 Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x001000 0x00000000 0x00000000 0x0c38a 0x0c38a R E 0x1000
  LOAD           0x00d38a 0x0000c38a 0x0000c38a 0x01506 0x01506 R   0x1000
  LOAD           0x00f000 0x0000e000 0x0000e000 0x0006c 0x0006c RW  0x1000
  LOAD           0x00f06c 0x0000e06c 0x0000e06c 0x0006c 0x0007c RW  0x1000
  DYNAMIC        0x00f000 0x0000e000 0x0000e000 0x00060 0x00060 RW  0x4
  GNU_RELRO      0x00f000 0x0000e000 0x0000e000 0x0006c 0x01000 R   0x1
  GNU_STACK      0x000000 0x00000000 0x00000000 0x00000 0x00000 RW  0
  EXIDX          0x00e444 0x0000d444 0x0000d444 0x002b0 0x002b0 R   0x4

 Section to Segment mapping:
  Segment Sections...
   00     .text
   01     .rodata .ARM.exidx .gnu.hash .ARM.extab .dynsym .dynstr .hash .rel.dyn
   02     .dynamic .got
   03     .data .bss
   04     .dynamic
   05     .dynamic .got .bss
   06
   07     .ARM.exidx

Nothing wrong with that from a strict ELF compliance point of view, but
it doesn't meet the requirements of our current gen_ldelf_hex.py script
which makes the build fail:

 $ scripts/gen_ldelf_hex.py --input out/arm-plat-vexpress/ldelf/ldelf.elf --output out/arm-plat-vexpress/core/ldelf_hex.c
 Expected load segment to be read/write

I think our script is a bit too strict, what really matters is that
OP-TEE creates two memory mappings for the PT_LOAD segments of ldelf,
one is RX and the other is RW. We can therefore concatenate segments as
long as we have one or more non-writable segments followed by one or
more writable ones.

This commit relaxes the requirements in gen_ldelf_hex.py and implements
the above conditions instead.

[1] clang version 11.0.0 (https://github.com/llvm/llvm-project.git
    6b3168f8cdb46656330929877b0b4daab35d30de)

Signed-off-by: Jerome Forissier <jerome@forissier.org>
```

### CODE REVIEW

**Comment by jforissier** [comment]
@vchong this fixes an issue I had with latest Clang from master (11.x) as I mentioned in the commit description. I did not had the same error as you, but you tried a different (older) commit.

File: scripts/gen_ldelf_hex.py [hunk 1]
```diff
@@ -48,16 +50,25 @@ def emit_load_segments(elffile, outf):
...
-                next_rwseg_va = segment['p_vaddr'] + segment['p_filesz']
+                if segment['p_flags'] & P_FLAGS.PF_W != P_FLAGS.PF_W:
+                    # A subsequent non-writable segment may be placed in the
+                    # "code" area (mapped RX)
+                    if data_size != 0:
+                        # Already met RW data: error
+                        print(f'Expected writable load segment (n={n})')
+                        sys.exit(1)
+                    else:
+                        if next_rseg_va and segment['p_vaddr'] != next_rseg_va:
```
Comment (jenswi-linaro) [inline scripts/gen_ldelf_hex.py:61]: This doesn't catch a gap between the RX and this R segment. Perhaps we can assign expected value above when taking care of the RX segment?

Comment (jforissier) [reply scripts/gen_ldelf_hex.py:61]: Good catch, I'll fix that

Comment (jforissier) [reply scripts/gen_ldelf_hex.py:61]: Hmmm... the problem is, Clang may also generate a gap ;-/ for example this has 4 byte "hole" between the first two LOAD segments:

```
Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x001000 0x00000000 0x00000000 0x04834 0x04834 R E 0x1000
  LOAD           0x005838 0x00004838 0x00004838 0x01620 0x01620 R   0x1000
```

I'll write `emit_load_segments()` a bit differently.

Comment (jenswi-linaro) [reply scripts/gen_ldelf_hex.py:61]: OK, some padding for proper alignment I suppose.

**Comment by jforissier** [comment]
Superseded by https://github.com/OP-TEE/optee_os/pull/3783.

**Pull request closed without merging**